File: src/main/java/no/mesan/fag/patterns/timesheet/format/StyleFactory.java
package no.mesan.fag.patterns.timesheet.format;

import java.util.HashMap;
import java.util.Map;

import org.apache.poi.ss.usermodel.CellStyle;
import org.apache.poi.ss.usermodel.Workbook;

/** Oppretter stiler for et worksheet. */
public final class StyleFactory {

    /** Tilgjengelige stiler. */
    public enum StyleName {
        H1, TBL_HEAD, TBL_HEAD_LEFT, COL1, COLN, SUMS, SUM1, DATA
    }

    /** Lag et stilbibliotek. */
    public static Map<StyleName, CellStyle> styleSetup(final Workbook wb) {
        return createWbStyles(wb, styleSetup());
    }

    private static Map<StyleName, StyleSpec> styleSetup(){
        final Map<StyleName, StyleSpec> styles = new HashMap<>();
        final BorderSpec.BorderLine mediumBorder = BorderSpec.BorderLine.MEDIUM_STD;
        final StyleSpec h1 = StyleSpec.newStyle()
                                      .bold().size(15).fgColor(ColorSpec.STD_H_COLOR).centerAlign().verticalCenter()
                                      .build();
        final StyleSpec tblHead = StyleSpec.newStyleFrom(h1).size(12).allBorders().rightAlign().build();
        final StyleSpec tblHeadLeft = StyleSpec.newStyleFrom(tblHead).italic().rightBorder(mediumBorder).build();
        final StyleSpec data = StyleSpec.newStyle().allBorders().build();
        final StyleSpec col1= StyleSpec.newStyleFrom(data).bold().leftAlign().rightBorder(mediumBorder).build();
        final StyleSpec colN= StyleSpec.newStyleFrom(data).italic().rightAlign().rightBorder(mediumBorder).build();
        final StyleSpec sums = StyleSpec.newStyle().italic().size(12).shaded().allBorders(mediumBorder).build();
        final StyleSpec sum1 = StyleSpec.newStyleFrom(sums).unitalic().bold().build();
        styles.put(StyleName.H1, h1);
        styles.put(StyleName.TBL_HEAD, tblHead);
        styles.put(StyleName.TBL_HEAD_LEFT, tblHeadLeft);
        styles.put(StyleName.COL1, col1);
        styles.put(StyleName.DATA, data);
        styles.put(StyleName.COLN, colN);
        styles.put(StyleName.SUMS, sums);
        styles.put(StyleName.SUM1, sum1);
        return styles;
    }

    private static Map<StyleName, CellStyle> createWbStyles(final Workbook wb, final Map<StyleName, StyleSpec> map){
        final Map<StyleName, CellStyle> styles = new HashMap<>();
        for (final Map.Entry<StyleName, StyleSpec> entry : map.entrySet()) {
            styles.put(entry.getKey(), entry.getValue().createStyle(wb));
        }
        return styles;
    }
}

File: src/main/scala/no/mesan/fag/patterns/scala/timesheet/format/StyleFactory.scala
package no.mesan.fag.patterns.scala.timesheet.format

import org.apache.poi.ss.usermodel.{CellStyle, Workbook}

/** Tilgjengelige stiler. */
abstract sealed class StyleName
case object H1 extends StyleName
case object TableHead extends StyleName
case object TableHeadLeft extends StyleName
case object Col1 extends StyleName
case object ColN extends StyleName
case object Sums extends StyleName
case object Sum1 extends StyleName
case object Data extends StyleName

/** Oppretter stiler for et worksheet. */
object StyleFactory {

  /** Lag et stilbibliotek. */
  def styleSetup(wb: Workbook): Map[StyleName, CellStyle] =  createWbStyles(wb, styleSetup)

  /// HINT Hvis Styles er endret som hintet, kan du
  //       1. fjerne new
  //       2. bruke navngitte argumenter i stedet for full argumentliste
  //       3. fjerne alle argumenter som har defaultverdi
  //       4. bruk den innebygde copy-metoden i case-klassene til å bygge en stil på en annen
  private val styleSetup: Map[StyleName, Styles] = Map(
    H1 -> new Styles(true, false, 15, Some(ColorHighlight), None, None, None, None, None,
      HorizontalCenter, VerticalMiddle),
    TableHead -> new Styles(true, false, 12, Some(ColorHighlight), Some(ColorBg), Some(BorderThin),
      Some(BorderThin), Some(BorderThin), Some(BorderThin), HorizontalRight, VerticalMiddle),
    TableHeadLeft -> new Styles(true, true, 12, Some(ColorHighlight), Some(ColorBg), Some(BorderThin),
      Some(BorderThin), Some(BorderThin), Some(BorderMedium), HorizontalLeft, VerticalMiddle),
    Col1 -> new Styles(true, false, 10, None, None, Some(BorderThin), Some(BorderThin),
      Some(BorderThin), Some(BorderMedium), HorizontalLeft, VerticalBottom),
    Data -> new Styles(false, false, 10, None, None, Some(BorderThin), Some(BorderThin),
      Some(BorderThin), Some(BorderThin), HorizontalGen, VerticalBottom),
    ColN -> new Styles(false, true, 10, None, None, Some(BorderThin), Some(BorderThin), Some(BorderThin),
      Some(BorderMedium), HorizontalRight, VerticalBottom),
    Sums -> new Styles(false, true, 12, Some(ColorShadeFg), Some(ColorShadeBg), Some(BorderMedium),
      Some(BorderMedium), Some(BorderMedium), Some(BorderMedium), HorizontalGen, VerticalBottom),
    Sum1 -> new Styles(true, false, 12, Some(ColorShadeFg), Some(ColorShadeBg), Some(BorderMedium),
      Some(BorderMedium), Some(BorderMedium), Some(BorderMedium), HorizontalGen, VerticalBottom)
  )

  private def createWbStyles(wb: Workbook, map: Map[StyleName, Styles]): Map[StyleName, CellStyle] =
    map.map { case (name, spec) => name -> spec.createStyle(wb) }
}
